src/main/java/edu/kit/ipd/crowdcontrol/objectservice/moneytransfer/MoneyTransferManager.java

```diff
@@ -0,0 +1,262 @@
+        for (Message message : messages) {
+            try {
+                GiftCodeRecord rec = MailParser.parseAmazonGiftCode(message);
+                workerBalanceOperations.addGiftCode(rec.getCode(), rec.getAmount());
+            } catch (MoneyTransferException e) {
+                try {
+                    mailHandler.markAsUnseen(message);
+                    throw new MoneyTransferException(e.getMessage());
+                } catch (MessagingException f) {
+                    f.printStackTrace();
+                }
+            }
+        }
+    }
+
+    private List<GiftCodeRecord> chooseGiftCodes(WorkerRecord worker, List<GiftCodeRecord> giftCodes) {
+        List<GiftCodeRecord> payedCodes = new ArrayList<>();
+        int creditBalance = workerBalanceOperations.getBalance(worker.getIdWorker());
+
+        for (GiftCodeRecord nextCode : giftCodes) {
+            if (creditBalance == 0) {
+                break;
+            }
+            if (nextCode.getAmount() <= creditBalance) {
+                payedCodes.add(nextCode);
+                creditBalance -= nextCode.getAmount();
+            }
+        }
+
+        if (creditBalance >= payOffThreshold && creditBalance >= 15 && !payedCodes.isEmpty()) {
+            notificationText = notificationText.append("A worker has pending Payments in the amount of ").append(creditBalance).append("ct. Please add giftcodes, so the payment of the worker can be continued.").append(System.getProperty("line.separator"));
+        }
+        return payedCodes;
+    }
+
+    private void payWorker(WorkerRecord worker, List<GiftCodeRecord> giftCodes) throws MoneyTransferException {
+        if (!giftCodes.isEmpty()) {
+            StringBuilder paymentMessage = loadMessage("src/main/resources/PaymentMessage.txt");
+            StringBuilder giftCodeMessage = new StringBuilder();
+
+            //saves payment to the database
+            for (GiftCodeRecord rec : giftCodes) {
+                workerBalanceOperations.addDebit(worker.getIdWorker(), rec.getAmount(), rec.getIdGiftCode());
+                giftCodeMessage.append(rec.getCode()).append(System.getProperty("line.separator"));
+            }
+
+            //creates payment message
+            Map<String, String> map = new HashMap<>();
+            map.put("GiftCodes", giftCodeMessage.toString());
+            paymentMessage = new StringBuilder(Template.apply(paymentMessage.toString(), map));
+
+            //sends payment message
+            try {
+                mailHandler.sendMail(worker.getEmail(), "Your payment for your Crowdworking", paymentMessage.toString());
+            } catch (MessagingException e) {
+                throw new MoneyTransferException("The MailHandler couldnt send mails to crowdworkers." +
+                        "It seems, that there is either a problem with the server or with the properties file.");
+            } catch (UnsupportedEncodingException e) {
+                LOGGER.error(e);
```

Comment:
For the cases that the messaging failed, shouldnt the gift code be recovered ? So he can be used again?